File: backend/app/schemas/team.py
"""팀 관련 Pydantic 요청/응답 스키마."""

from __future__ import annotations

from datetime import datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field


# ── Response 스키마 ──


class TeamMemberInfo(BaseModel):
    id: int
    team_id: str
    role: str
    nickname: str
    description: str | None = None
    system_prompt: str | None = None
    allowed_tools: str | None = None
    disallowed_tools: str | None = None
    model: str | None = None
    max_turns: int | None = None
    max_budget_usd: float | None = None
    mcp_server_ids: list[str] | None = None
    created_at: datetime
    updated_at: datetime


class TaskSummary(BaseModel):
    total: int = 0
    pending: int = 0
    in_progress: int = 0
    completed: int = 0
    failed: int = 0


class TeamInfo(BaseModel):
    id: str
    name: str
    description: str | None = None
    status: str
    lead_member_id: int | None = None
    config: dict | None = None
    created_at: datetime
    updated_at: datetime
    members: list[TeamMemberInfo] = []
    task_summary: TaskSummary = TaskSummary()


class TeamListItem(BaseModel):
    id: str
    name: str
    description: str | None = None
    status: str
    lead_member_id: int | None = None
    created_at: datetime
    updated_at: datetime
    member_count: int = 0
    task_summary: TaskSummary = TaskSummary()


# ── Request 스키마 ──


class CreateTeamRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=100)
    description: Optional[str] = None
    config: Optional[dict] = None


class UpdateTeamRequest(BaseModel):
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = None
    status: Optional[Literal["active", "completed", "paused", "archived"]] = None
    config: Optional[dict] = None


class AddTeamMemberRequest(BaseModel):
    nickname: str = Field(..., min_length=1, max_length=50)
    role: Literal["lead", "member"] = "member"
    description: Optional[str] = None
    system_prompt: Optional[str] = None
    allowed_tools: Optional[str] = None
    disallowed_tools: Optional[str] = None
    model: Optional[str] = None
    max_turns: Optional[int] = None
    max_budget_usd: Optional[float] = None
    mcp_server_ids: Optional[list[str]] = None


class UpdateTeamMemberRequest(BaseModel):
    nickname: Optional[str] = Field(None, min_length=1, max_length=50)
    role: Optional[Literal["lead", "member"]] = None
    description: Optional[str] = None
    system_prompt: Optional[str] = None
    allowed_tools: Optional[str] = None
    disallowed_tools: Optional[str] = None
    model: Optional[str] = None
    max_turns: Optional[int] = None
    max_budget_usd: Optional[float] = None
    mcp_server_ids: Optional[list[str]] = None


class SetLeadRequest(BaseModel):
    member_id: int


# ── 태스크 스키마 ──


class TeamTaskInfo(BaseModel):
    id: int
    team_id: str
    title: str
    description: str | None = None
    status: str
    priority: str
    assigned_member_id: int | None = None
    assigned_nickname: str | None = None
    created_by_member_id: int | None = None
    work_dir: str
    session_id: str | None = None
    result_summary: str | None = None
    order_index: int = 0
    depends_on_task_id: int | None = None
    created_at: datetime
    updated_at: datetime


class CreateTaskRequest(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: Optional[str] = None
    priority: Literal["low", "medium", "high"] = "medium"
    work_dir: str = Field(..., min_length=1)
    assigned_member_id: Optional[int] = None
    depends_on_task_id: Optional[int] = None


class UpdateTaskRequest(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=200)
    description: Optional[str] = None
    status: Optional[
        Literal["pending", "in_progress", "completed", "failed", "cancelled"]
    ] = None
    priority: Optional[Literal["low", "medium", "high"]] = None
    work_dir: Optional[str] = None
    assigned_member_id: Optional[int] = None
    order_index: Optional[int] = None


class ReorderTasksRequest(BaseModel):
    task_ids: list[int] = Field(..., min_length=1)


class CompleteTaskRequest(BaseModel):
    result_summary: Optional[str] = None


class DelegateTaskRequest(BaseModel):
    member_id: Optional[int] = None
    prompt: Optional[str] = None


# ── 메시지 스키마 ──


class TeamMessageInfo(BaseModel):
    id: int
    team_id: str
    from_member_id: int
    to_member_id: int | None = None
    content: str
    message_type: str
    metadata_json: str | None = None
    is_read: bool = False
    created_at: datetime
    from_nickname: str | None = None


class SendMessageRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=5000)
    from_member_id: int
    to_member_id: Optional[int] = None
    message_type: Literal["info", "task_update", "request", "result", "delegate"] = (
        "info"
    )
    metadata_json: Optional[str] = None


class MarkReadRequest(BaseModel):
    message_ids: list[int] = Field(..., min_length=1)

File: frontend/src/types/team.ts
/**
 * 팀 관련 타입 정의.
 */

export interface TeamMemberInfo {
  id: number;
  team_id: string;
  role: "lead" | "member";
  nickname: string;
  description: string | null;
  system_prompt: string | null;
  allowed_tools: string | null;
  disallowed_tools: string | null;
  model: string | null;
  max_turns: number | null;
  max_budget_usd: number | null;
  mcp_server_ids: string[] | null;
  created_at: string;
  updated_at: string | null;
}

export interface TaskSummary {
  total: number;
  pending: number;
  in_progress: number;
  completed: number;
  failed: number;
}

export interface TeamInfo {
  id: string;
  name: string;
  description: string | null;
  status: "active" | "completed" | "paused" | "archived";
  lead_member_id: number | null;
  config: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
  members: TeamMemberInfo[];
  task_summary: TaskSummary;
}

export interface TeamListItem {
  id: string;
  name: string;
  description: string | null;
  status: "active" | "completed" | "paused" | "archived";
  lead_member_id: number | null;
  created_at: string;
  updated_at: string;
  member_count: number;
  task_summary: TaskSummary;
}

export interface CreateTeamRequest {
  name: string;
  description?: string;
  config?: Record<string, unknown>;
}

export interface UpdateTeamRequest {
  name?: string;
  description?: string;
  status?: "active" | "completed" | "paused" | "archived";
  config?: Record<string, unknown>;
}

export interface AddTeamMemberRequest {
  nickname: string;
  role?: "lead" | "member";
  description?: string;
  system_prompt?: string;
  allowed_tools?: string;
  disallowed_tools?: string;
  model?: string;
  max_turns?: number;
  max_budget_usd?: number;
  mcp_server_ids?: string[];
}

export interface UpdateTeamMemberRequest {
  nickname?: string;
  role?: "lead" | "member";
  description?: string;
  system_prompt?: string;
  allowed_tools?: string;
  disallowed_tools?: string;
  model?: string;
  max_turns?: number;
  max_budget_usd?: number;
  mcp_server_ids?: string[];
}

export interface SetLeadRequest {
  member_id: number;
}

// ── 태스크 ──

export type TaskStatus = "pending" | "in_progress" | "completed" | "failed" | "cancelled";
export type TaskPriority = "low" | "medium" | "high";

export interface TeamTaskInfo {
  id: number;
  team_id: string;
  title: string;
  description: string | null;
  status: TaskStatus;
  priority: TaskPriority;
  assigned_member_id: number | null;
  assigned_nickname: string | null;
  created_by_member_id: number | null;
  workspace_id: string | null;
  workspace_name: string | null;
  session_id: string | null;
  result_summary: string | null;
  order_index: number;
  depends_on_task_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface CreateTaskRequest {
  title: string;
  description?: string;
  priority?: TaskPriority;
  workspace_id: string;
  assigned_member_id?: number;
  depends_on_task_id?: number;
}

export interface UpdateTaskRequest {
  title?: string;
  description?: string;
  status?: TaskStatus;
  priority?: TaskPriority;
  workspace_id?: string;
  assigned_member_id?: number;
  order_index?: number;
}

export interface CompleteTaskRequest {
  result_summary?: string;
}

export interface DelegateTaskRequest {
  member_id?: number;
  prompt?: string;
}

// ── 메시지 ──

export type TeamMessageType = "info" | "task_update" | "request" | "result" | "delegate";

export interface TeamMessageInfo {
  id: number;
  team_id: string;
  from_member_id: number;
  to_member_id: number | null;
  content: string;
  message_type: TeamMessageType;
  metadata_json: string | null;
  is_read: boolean;
  created_at: string;
  from_nickname: string | null;
}

export interface SendMessageRequest {
  from_member_id: number;
  content: string;
  to_member_id?: number;
  message_type?: TeamMessageType;
  metadata_json?: string;
}

export interface MarkReadRequest {
  message_ids: number[];
}
